### PR TITLE
feat: support react-aria-components render prop

### LINKED ISSUE #11: Customize `className` with props from underlying component `className` render function.
## 🚀 Feature Proposal

Some components from `react-aria-components` accept a function as `className` prop value. The function will receive the render props. So, it basically a render function for `className`. `react-twc` supports customizing `className` via props, might as well merge `className` render props from such components with the transient props. I use this pattern a lot while wrapping `react-aria-components`.

## Motivation

To properly customize the produced `className` using the underlying `className` render props.

## Example

Without `react-twc`:
```javascript
import * as React from 'react'
import * as RAC from 'react-aria-components'
import { twMerge as tw } from 'tailwind-merge'

const Button = React.forwardRef(function Button({ className, ...props }, ref) {
    return ( 
        <RAC.Button 
            {...props} 
            className={(renderProps) => {
                return tw('flex px-2', typeof className === 'function' ? className(renderProps) : className)
            }
        } />
    )
})
```

With `react-twc`:
```javascript
import * as RAC from 'react-aria-components'
import { twc } from 'react-twc'

const Button = twc(RAC.Button)(
    (props) => ['flex px-2', props.isHovered ? 'bg-blue-700' : 'bg-blue-600'], { asFunction: true }
)
```

## Pitch

By supporting `classname` render function, `react-twc` will be fully compatible with component library such as `react-aria-components` and `headlessui`.

### PR DESCRIPTION
**Pull request by gregberge**
Fix #11

# Render props

Some libraries like [`react-aria-components`](https://react-spectrum.adobe.com/react-aria/index.html) or [Remix](https://remix.run/docs/en/main/components/nav-link) accepts a function as `className`. `tsc` supports render props out of the box.

## Usage

In this example, we use a render props to change the class of a [`react-aria-components` button](https://react-spectrum.adobe.com/react-aria/Button.html) when the button is pressed.

```tsx {5,6}
import { twc } from 'react-twc'
import { Button as AriaButton, ButtonProps } from "react-aria-components";

const Button = twc(AriaButton)<ButtonProps>(
  (props) => (({ isPressed })) =>
    isPressed ? "bg-gray-700" : "bg-gray-500",
);

export default () => (
  <Button>Click me</Button>
);
```

### CODE REVIEW

File: src/index.test.tsx [hunk 1]
```diff
@@ -199,4 +204,26 @@ describe("twc", () => {
...
     expect(title.classList.contains("font-bold")).toBe(false);
   });
+
+  test("works with render props", () => {
+    const twx = createTwc({
+      compose: (fn, className) =>
+        typeof fn === "function"
+          ? composeRenderProps(className, (_prev, renderProps) =>
+              fn(renderProps),
+            )
```
Comment (izznatsir) [inline src/index.test.tsx:214]: `_prev` should be composed with the return value of `fp(renderProps`. Wrap it with `twMerge` or `clsx` will suffice.

```javascript
...
        ? composeRenderProps(className, (prevClassName, renderProps) =>
                twMerge(prevClassName, fn(renderProps)),
            )
...
```

**Review by izznatsir — CHANGES_REQUESTED** [review]

**Comment by izznatsir** [comment]
This approach is brilliant!

**Comment by gregberge** [comment]
@devongovett could you please give a review on this? What do you think?

https://twc-o3f01nfwe-argos-ci.vercel.app/docs/integrations/react-aria-components

**Comment by gregberge** [comment]
@izznatsir final implementation, please review :)

**Comment by gregberge** [comment]
I finally find a way to support it without any adapter 🎉

**Comment by ducan-ne** [comment]
@gregberge Do you think this API could be way better?
```js
const Button = twc(AriaButton)<ButtonProps>(
  (props, { isPressed }) => =>
    isPressed ? "bg-gray-700" : "bg-gray-500",
);
```
I'm uncertain about the technical aspect.

**Pull request merged**